### PR TITLE
Yukarinn commands

### PR DESCRIPTION
**Pull request by Yukarinn**

### CODE REVIEW

File: dmoj/cli.py [hunk 1]
```diff
@@ -310,6 +317,83 @@ def execute(self, line):
...
+
+        return src, err
+
+    def execute(self, line):
+        args = self.arg_parser.parse_args(line)
+
+        data1, err1 = self.get_data(args.id_or_source_1)
+        data2, err2 = self.get_data(args.id_or_source_2)
+
+        file_diff = '\n'.join(list(difflib.unified_diff(data1.splitlines(), data2.splitlines(), fromfile=args.id_or_source_1, tofile=args.id_or_source_2, lineterm='')))
```
Comment (quantum5) [inline dmoj/cli.py]: Can you use shorter lines?

File: dmoj/cli.py [hunk 2]
```diff
@@ -310,6 +351,87 @@ def execute(self, line):
...
+        self.arg_parser.add_argument('id_or_source', help='id or path of submission to show', metavar='<source>')
+
+    def get_data(self, id_or_source):
+        err = None
+        src = None
+        lexer = None
+        try:
+            id = int(id_or_source)
+            try:
+                global graded_submissions
```
Comment (quantum5) [inline dmoj/cli.py]: What the hell is this? Why do you need `global`?

File: dmoj/cli.py [hunk 3]
```diff
@@ -310,6 +351,87 @@ def execute(self, line):
...
+        self.arg_parser.add_argument('id_or_source_1', help='id or path of first source', metavar='<source 1>')
+        self.arg_parser.add_argument('id_or_source_2', help='id or path of second source', metavar='<source 2>')
+
+    def get_data(self, id_or_source):
+        err = None
+        src = None
+        try:
+            id = int(id_or_source)
+            try:
+                global graded_submissions
```
Comment (quantum5) [inline dmoj/cli.py]: What the hell.

File: dmoj/cli.py [hunk 4]
```diff
@@ -267,6 +296,18 @@ def execute(self, line):
...
             return
 
+        try:
+            with tempfile.NamedTemporaryFile() as temp:
+                temp.write('%s' % src)
+                temp.flush()
+                subprocess.call(['nano', temp.name])
+                temp.seek(0)
+                src = temp.read()
+                temp.close()
```
Comment (quantum5) [inline dmoj/cli.py]: It should be automatically closed due to the context manager.

File: dmoj/cli.py [hunk 5]
```diff
@@ -267,6 +296,18 @@ def execute(self, line):
             print ansi_style('#ansi[%s](red|bold)\n' % err)
             return
 
+        try:
+            with tempfile.NamedTemporaryFile() as temp:
+                temp.write('%s' % src)
+                temp.flush()
+                subprocess.call(['nano', temp.name])
```
Comment (quantum5) [inline dmoj/cli.py]: Remember it's specified to use `$EDITOR`.

File: dmoj/cli.py [hunk 6]
```diff
@@ -267,6 +296,18 @@ def execute(self, line):
...
+                temp.seek(0)
+                src = temp.read()
+                temp.close()
+        except Exception:
+            for warning in judgeenv.startup_warnings:
+                print ansi_style('#ansi[Warning: %s](yellow)' % warning)
```
Comment (quantum5) [inline dmoj/cli.py]: Why the hell are you printing all the existing startup warnings here?

File: dmoj/cli.py [hunk 7]
```diff
@@ -202,17 +231,17 @@ def execute(self, line):
...
-                        s = raw_input()
-                        if s.strip() == ':q':
-                            raise EOFError
-                        src.append(s)
-                except EOFError:  # Ctrl+D
-                    src = '\n'.join(src)
-                except Exception as io:
-                    err = str(io)
+                    with tempfile.NamedTemporaryFile(suffix=file_suffix) as temp:
+                        subprocess.call(['nano', temp.name])
```
Comment (quantum5) [inline dmoj/cli.py]: Aren't you supposed to keep the old code around for when `$EDITOR` is not defined?

Comment (quantum5) [inline dmoj/cli.py]: And also, it said to use `$EDITOR`...

File: dmoj/cli.py [hunk 8]
```diff
@@ -267,6 +296,18 @@ def execute(self, line):
             print ansi_style('#ansi[%s](red|bold)\n' % err)
             return
 
+        try:
+            with tempfile.NamedTemporaryFile() as temp:
```
Comment (quantum5) [inline dmoj/cli.py]: What happened to the extension?

File: requirements.txt [hunk 9]
```diff
@@ -4,3 +4,5 @@ ansi2html
 termcolor
 colorama
 cython
+pykwalify
```
Comment (quantum5) [inline requirements.txt]: As far as I can tell, `pykwalify` is not a requirement for this pull request.

File: setup.py [hunk 10]
```diff
@@ -97,7 +97,7 @@ def unavailable(self, e):
         ]
     },
     ext_modules=cythonize(extensions),
-    install_requires=['watchdog', 'pyyaml', 'ansi2html', 'termcolor'],
+    install_requires=['watchdog', 'pyyaml', 'ansi2html', 'termcolor', 'pykwalify', 'pygments'],
```
Comment (quantum5) [inline setup.py]: As far as I can tell, `pykwalify` is not a requirement for this pull request.

File: dmoj/cli.py [hunk 11]
```diff
@@ -181,13 +188,35 @@ def execute(self, line):
...
-        language_id = args.language_id
         time_limit = args.time_limit
         memory_limit = args.memory_limit
+        language_id = args.language_id
 
         err = None
+
         if problem_id not in map(itemgetter(0), judgeenv.get_supported_problems()):
             err = "unknown problem '%s'" % problem_id
+        elif language_id == ' ':
```
Comment (quantum5) [inline dmoj/cli.py]: What the hell. Why would you use a goddamn space as the sentinel value???

File: dmoj/cli.py [hunk 12]
```diff
@@ -168,8 +175,8 @@ class SubmitCommand(Command):
 
     def _populate_parser(self):
         self.arg_parser.add_argument('problem_id', help='id of problem to grade')
-        self.arg_parser.add_argument('language_id', help='id of the language to grade in (e.g., PY2)')
         self.arg_parser.add_argument('source_file', nargs='?', help='path to submission source (optional)')
+        self.arg_parser.add_argument('-lan', '--language-id', type=str, default=' ', help='id of the language to grade in (e.g., PY2)')
```
Comment (quantum5) [inline dmoj/cli.py]: That space...

File: dmoj/cli.py [hunk 13]
```diff
@@ -202,17 +231,19 @@ def execute(self, line):
                 except Exception as io:
                     err = str(io)
             else:
-                src = []
+                file_suffix = '.' + language_id
```
Comment (quantum5) [inline dmoj/cli.py]: This is merely ridiculous. You think `.PY2` is totally a valid extension? Not just `.py2`, but capital `.PY2`.

File: dmoj/cli.py [hunk 14]
```diff
@@ -202,17 +231,19 @@ def execute(self, line):
...
                 try:
-                    while True:
-                        s = raw_input()
-                        if s.strip() == ':q':
-                            raise EOFError
-                        src.append(s)
-                except EOFError:  # Ctrl+D
-                    src = '\n'.join(src)
+                    EDITOR = os.environ.get('EDITOR')
```
Comment (quantum5) [inline dmoj/cli.py]: Why the hell would you use capital variable names????

File: dmoj/cli.py [hunk 15]
```diff
@@ -267,6 +298,17 @@ def execute(self, line):
             print ansi_style('#ansi[%s](red|bold)\n' % err)
             return
 
+        try:
+            file_suffix = '.' + lang
```
Comment (quantum5) [inline dmoj/cli.py]: This extension...

File: dmoj/cli.py [hunk 16]
```diff
@@ -267,6 +298,17 @@ def execute(self, line):
...
+            with tempfile.NamedTemporaryFile(suffix=file_suffix) as temp:
+                temp.write('%s' % src)
```
Comment (quantum5) [inline dmoj/cli.py]: What's this brain damage? What's so hard about writing the string directly?

File: dmoj/cli.py [hunk 17]
```diff
@@ -267,6 +298,17 @@ def execute(self, line):
...
+                temp.flush()
+                subprocess.call(['nano', temp.name])
```
Comment (quantum5) [inline dmoj/cli.py]: Why are you *still* hardcoding `nano` here‽‽‽‽‽

File: dmoj/cli.py [hunk 18]
```diff
@@ -267,6 +298,17 @@ def execute(self, line):
...
+                temp.seek(0)
+                src = temp.read()
+        except Exception as io:
```
Comment (quantum5) [inline dmoj/cli.py]: What kind of exception variable name is `io`?

File: dmoj/cli.py [hunk 19]
```diff
@@ -310,6 +352,85 @@ def execute(self, line):
...
+    def _populate_parser(self):
+        self.arg_parser.add_argument('id_or_source_1', help='id or path of first source', metavar='<source 1>')
+        self.arg_parser.add_argument('id_or_source_2', help='id or path of second source', metavar='<source 2>')
+
+    def get_data(self, id_or_source):
+        err = None
+        src = None
+        try:
+            id = int(id_or_source)
+            global graded_submissions
```
Comment (quantum5) [inline dmoj/cli.py]: Why the hell are you still doing `global`?

File: dmoj/cli.py [hunk 20]
```diff
@@ -310,6 +352,85 @@ def execute(self, line):
...
+            problem, lang, src, tl, ml = graded_submissions[id - 1]
+        except IndexError:
+            err = "invalid submission '%d'" % (id - 1)
+        except ValueError:
+            try:
+                with open(os.path.realpath(id_or_source), 'r') as f:
+                    src = f.read()
+            except Exception as io:
```
Comment (quantum5) [inline dmoj/cli.py]: What kind of name is `io`?

File: dmoj/cli.py [hunk 21]
```diff
@@ -310,6 +352,85 @@ def execute(self, line):
...
+        self.arg_parser.add_argument('id_or_source_2', help='id or path of second source', metavar='<source 2>')
+
+    def get_data(self, id_or_source):
+        err = None
+        src = None
+        try:
+            id = int(id_or_source)
+            global graded_submissions
+            problem, lang, src, tl, ml = graded_submissions[id - 1]
+        except IndexError:
```
Comment (quantum5) [inline dmoj/cli.py]: Put the `id = int(id_or_source)` into a different try block so you don't end up with some retarded control flow.

File: dmoj/cli.py [hunk 22]
```diff
@@ -310,6 +352,85 @@ def execute(self, line):
...
+
+    def _populate_parser(self):
+        self.arg_parser.add_argument('id_or_source', help='id or path of submission to show', metavar='<source>')
+
+    def get_data(self, id_or_source):
+        err = None
+        src = None
+        lexer = None
+        try:
+            global graded_submissions
```
Comment (quantum5) [inline dmoj/cli.py]: Why is this `global` brain damage still here?

File: dmoj/cli.py [hunk 23]
```diff
@@ -181,13 +188,35 @@ def execute(self, line):
...
-        language_id = args.language_id
         time_limit = args.time_limit
         memory_limit = args.memory_limit
+        language_id = args.language_id
 
         err = None
+
         if problem_id not in map(itemgetter(0), judgeenv.get_supported_problems()):
             err = "unknown problem '%s'" % problem_id
+        elif language_id == 'unknown':
```
Comment (quantum5) [inline dmoj/cli.py]: I told you to use `None`, goddamnit. I TOLD YOU AND YOU WON"T LISTEN.

File: dmoj/cli.py [hunk 24]
```diff
@@ -181,13 +188,35 @@ def execute(self, line):
...
+            if args.source_file:
+                ext = None
+                try:
+                    filename, ext = args.source_file.split('.')
+                except Exception:
+                    err = 'invalid file name'
+                ext = ext.upper()
+                print ext
```
Comment (quantum5) [inline dmoj/cli.py]: What the fuck.

File: dmoj/cli.py [hunk 25]
```diff
@@ -202,17 +231,34 @@ def execute(self, line):
...
-                    while True:
-                        s = raw_input()
-                        if s.strip() == ':q':
-                            raise EOFError
-                        src.append(s)
-                except EOFError:  # Ctrl+D
-                    src = '\n'.join(src)
-                except Exception as io:
-                    err = str(io)
+                if language_id in ['PY2', 'PYPY2']:
```
Comment (quantum5) [inline dmoj/cli.py]: Are you so brain damaged as to not realize there is `.ext` attribute on the executor? How the hell did you think executors know how to create source files?? Stop reinventing the goddamn wheel in a shitty way.

File: dmoj/cli.py [hunk 26]
```diff
@@ -202,17 +231,34 @@ def execute(self, line):
...
+                else:
+                    file_suffix = '.' + language_id.lower()
+
+                editor = os.environ.get('EDITOR')
+                if editor:
+                    with tempfile.NamedTemporaryFile(suffix= file_suffix) as temp:
+                        subprocess.call([editor, temp.name])
+                        temp.seek(0)
+                        src = temp.read()
+                else:
```
Comment (quantum5) [inline dmoj/cli.py]: What happened to the warning?

File: dmoj/cli.py [hunk 27]
```diff
@@ -267,6 +313,36 @@ def execute(self, line):
             print ansi_style('#ansi[%s](red|bold)\n' % err)
             return
 
+        if lang in ['PY2', 'PYPY2']:
```
Comment (quantum5) [inline dmoj/cli.py]: Once is bad enough, goddamnit.

File: dmoj/cli.py [hunk 28]
```diff
@@ -267,6 +313,36 @@ def execute(self, line):
...
+
+        editor = os.environ.get('EDITOR')
+        if editor:
+            with tempfile.NamedTemporaryFile(suffix=file_suffix) as temp:
+                temp.write(src)
+                temp.flush()
+                subprocess.call([editor, temp.name])
+                temp.seek(0)
+                src = temp.read()
+        else:
```
Comment (quantum5) [inline dmoj/cli.py]: Why don't you put this into a function so you don't duplicate this?

File: dmoj/cli.py [hunk 29]
```diff
@@ -310,6 +386,86 @@ def execute(self, line):
...
+                    lexer = pygments.lexers.PythonLexer()
+                elif lang in ['PY3', 'PYPY3']:
+                    lexer = pygments.lexers.Python3Lexer()
+                else:
+                    lexer = pygments.lexers.guess_lexer(src)
+            except IndexError:
+                err = "invalid submission '%d'" % (id - 1)
+        except ValueError:
+            try:
+                with open(os.path.realpath(id_or_source), 'r') as f:
```
Comment (quantum5) [inline dmoj/cli.py]: `'r'` is the default mode...

File: dmoj/cli.py [hunk 30]
```diff
@@ -181,13 +188,35 @@ def execute(self, line):
...
         err = None
+
         if problem_id not in map(itemgetter(0), judgeenv.get_supported_problems()):
             err = "unknown problem '%s'" % problem_id
+        elif language_id == 'unknown':
+            if args.source_file:
+                ext = None
+                try:
+                    filename, ext = args.source_file.split('.')
+                except Exception:
```
Comment (quantum5) [inline dmoj/cli.py]: What the hell. Why don't you just use `partition`?

File: dmoj/cli.py [hunk 31]
```diff
@@ -168,8 +175,8 @@ class SubmitCommand(Command):
 
     def _populate_parser(self):
         self.arg_parser.add_argument('problem_id', help='id of problem to grade')
-        self.arg_parser.add_argument('language_id', help='id of the language to grade in (e.g., PY2)')
         self.arg_parser.add_argument('source_file', nargs='?', help='path to submission source (optional)')
+        self.arg_parser.add_argument('-lan', '--language-id', type=str, default='unknown', help='id of the language to grade in (e.g., PY2)')
```
Comment (quantum5) [inline dmoj/cli.py]: I don't know about you, but I certainly don't want to type `submit <problem> -lan PY2`. I want to type in `submit <problem> PY2` like before to type my code into `$EDITOR`. Not everyone is as masochistic as you are.

File: dmoj/cli.py [hunk 32]
```diff
@@ -188,33 +188,36 @@ def execute(self, line):
...
-                    language_id = ext
                 else:
-                    err = "unknown language '%s'" % language_id
+                    ext = ext.upper()
+                    if ext == 'PY':
+                        language_id = 'PY2'
+                    elif ext == 'CPP':
+                        language_id = 'CPP14'
+                    elif ext == 'JAVA':
+                        language_id = 'JAVA9'
```
Comment (Xyene) [inline dmoj/cli.py]: I'd go with JAVA8 here

File: dmoj/cli.py [hunk 33]
```diff
@@ -188,33 +188,36 @@ def execute(self, line):
...
-                    language_id = 'JAVA9'
-                elif ext:
-                    language_id = ext
                 else:
-                    err = "unknown language '%s'" % language_id
+                    ext = ext.upper()
+                    if ext == 'PY':
+                        language_id = 'PY2'
+                    elif ext == 'CPP':
+                        language_id = 'CPP14'
```
Comment (Xyene) [inline dmoj/cli.py]: CPP11 here.

In theory we ought to have this use the latest version of CPP available, and the latest version of Java available etc., but in practice I doubt that'd be worth the extra code complexity (this tool is intended mostly for developers, who likely know what they're doing).

**Pull request merged**